crypto/bio/bio_lib.c

```diff
@@ -786,55 +786,66 @@ void bio_cleanup(void)
 }
 
 /* Internal variant of the below BIO_wait() not calling BIOerr() */
-static int bio_wait(BIO *bio, time_t max_time, unsigned int milliseconds)
+static int bio_wait(BIO *bio, time_t max_time, unsigned int nap_milliseconds)
 {
 #ifndef OPENSSL_NO_SOCK
     int fd;
 #endif
+    long sec_diff;
 
-    if (max_time == 0)
+    if (max_time == 0) /* no timeout */
         return 1;
 
 #ifndef OPENSSL_NO_SOCK
-    if (BIO_get_fd(bio, &fd) > 0)
+    if (BIO_get_fd(bio, &fd) > 0 && fd < FD_SETSIZE)
         return BIO_socket_wait(fd, BIO_should_read(bio), max_time);
 #endif
-    if (milliseconds > 1000) {
-        long sec_diff = (long)(max_time - time(NULL)); /* might overflow */
+    /* fall back to polling since no sockets are available */
 
-        if (sec_diff <= 0)
-            return 0; /* timeout */
-        if ((unsigned long)sec_diff < milliseconds / 1000)
-            milliseconds = (unsigned long)sec_diff * 1000;
+    sec_diff = (long)(max_time - time(NULL)); /* might overflow */
+    if (sec_diff < 0)
+        return 0; /* clearly timeout */
+
+    /* now take a nap at most the given number of milliseconds */
+    if (sec_diff == 0) { /* we are below the 1 seconds resolution of max_time */
+        if (nap_milliseconds > 1000)
+            nap_milliseconds = 1000;
+    } else { /* for sec_diff > 0, take min(sec_diff * 1000, nap_milliseconds) */
+        if ((unsigned long)sec_diff * 1000 < nap_milliseconds)
+            nap_milliseconds = (unsigned int)sec_diff * 1000;
     }
-    ossl_sleep(milliseconds);
+    ossl_sleep(nap_milliseconds);
     return 1;
 }
 
-/*
+/*-
  * Wait on (typically socket-based) BIO at most until max_time.
- * Succeed immediately if max_time == 0. If sockets are not available succeed
- * after waiting at most given milliseconds in order to avoid a tight busy loop.
- * Call BIOerr(...) unless success.
+ * Succeed immediately if max_time == 0.
+ * If sockets are not available support polling: succeed after waiting at most
+ * the number of nap_milliseconds in order to avoid a tight busy loop.
+ * Call BIOerr(...) on timeout or error.
  * Returns -1 on error, 0 on timeout, and 1 on success.
  */
-int BIO_wait(BIO *bio, time_t max_time, unsigned int milliseconds)
+int BIO_wait(BIO *bio, time_t max_time, unsigned int nap_milliseconds)
 {
-    int rv = bio_wait(bio, max_time, milliseconds);
+    int rv = bio_wait(bio, max_time, nap_milliseconds);
 
     if (rv <= 0)
         BIOerr(0, rv == 0 ? BIO_R_TRANSFER_TIMEOUT : BIO_R_TRANSFER_ERROR);
     return rv;
 }
 
 /*
- * Connect via given BIO using BIO_do_handshake() until success/timeout/error.
- * Parameter timeout == 0 means infinite, < 0 leads to immediate timeout error.
+ * Connect via given BIO using BIO_do_connect() until success/timeout/error.
+ * Parameter timeout == 0 means no timeout, < 0 means exactly one try.
+ * For non-blocking and potentially even non-socket BIOs perform polling with
```

Comment:
which non-socket BIOs have you in mind?
have you tested this code with one of those?

Reply:
I don't have such a non-socket BIO at hand, so could not test this.
That's part of the reason I wrote "potentially even non-socket BIOs".
Anyway, for non-blocking socket-based BIOs (i.e., timeout > 0) polling is needed.

Reply:
Well, that's why I'd initially restrict the scope of such a function,
to the domain where it is used, and extend that later
when other non-socket use become available.
I have the impression that must be invented first.